ARCHITECTURE.md

```diff
@@ -0,0 +1,74 @@
+# Exonum Architecture
+
+This document describes the high-level architecture of Exonum. Its target
+audience is the developers of Exonum core. To learn how to *use* Exonum, see
+https://exonum.com/doc/.
+
+This document is provided on the best effort bases -- it is not guaranteed to be
+up-to-date and complete. The purpose of this document is to highlight the most
+important types, modules and functions in the code, to allow new contributors to
+dive into Exonum quickly. It is assumed that the reader is familiar with core
+concepts of Exonum, such as services and transactions.
+
+
+## Code organization
+
+The repository is a single Cargo workspace of several related packages:
```

Comment:
There's no right answer here really because the terminology used is contradictory :( However, I personally would prefer a `package` here. For the reference, the current mess looks like this:

- `crate` is, formally, a compilation unit. It is a tree of modules, which starts at `lib.rs` / `main.rs` / `example/foo.rs` and includes other files via `mod foo` declarations. `crate` is a term, specific to the Rust language. 

- `package` is "cohesive unit of distribution" of rust code. It is a term, specific to Cargo. A package typically consists of several crates: the library in `src/lib.rs`, the binary in `src/main.rs`, the integration tests and examples. The `package` terminology is used in `Cargo.toml` (it begins with a `[package]` section) and in certain flags (`cargo test --package exonum`).

- `target`: confusingly, this is how Cargo calls crates which constitute a package. I.e., there's a library target, a binary target, and `--all-targets` flag.

- `target-tripple`: this specifies the identity of architecture for cross-compiling. It is often referred to as simply "target".

- `crate`: confusingly, this is how the `crates.io` website calls the `package`. 

The Rust book also seems to use package: https://doc.rust-lang.org/book/second-edition/ch14-03-cargo-workspaces.html